tools/extra/fpgadiag/mactest.py

```diff
@@ -0,0 +1,171 @@
+                with open(os.path.join(SYS_IF, i, 'address')) as f:
+                    self.ethif[i] = f.read().strip()
+
+        if self.ethif:
+            print('Found {} ethernet interfaces:'.format(len(self.ethif)))
+            ethifs = sorted(self.ethif.items())
+            for i in ethifs:
+                print('  {:<20} {}'.format(*i))
+            print(DIVIDE)
+        else:
+            exception_quit('No ethernet interface found!')
+
+    def str2hex(self, c):
+        return int(binascii.b2a_hex(c), 16)
+
+    def read_mac_from_nvmem(self):
+        mac_number = 0
+        mac = 0
+        with open(self.args.nvmem, 'rb') as f:
+            f.seek(self.args.offset, 0)
+            for x in [40, 32, 24, 16, 8, 0]:
+                mac |= self.str2hex(f.read(1)) << x
+            mac_number = self.str2hex(f.read(1))
+        print('Read {} mac addresses from NVMEM:'.format(mac_number))
+        vendor = '{:06x}'.format(mac >> 24)
+        for i in range(mac_number):
+            mac += 1
+            mac_str = '{}{:06x}'.format(vendor, (mac & 0xffffff))
+            fmt_str = ':'.join([mac_str[x:x + 2]
+                                for x in range(0, len(mac_str), 2)])
+            self.mac.append(fmt_str)
+        for m in self.mac:
+            print('  {}'.format(m))
+        print(DIVIDE)
+
+    def compare_eth_mac_with_macrom(self):
+        result = 'PASS'
+        for m in self.mac:
+            if m not in self.ethif.values():
+                print('{} in MAC ROM, have no relative interface!'.format(m))
+                result = 'FAIL'
+        ethifs = sorted(self.ethif.items())
+        for e, m in ethifs:
+            if m not in self.mac:
+                print('{} mac {}, is not from MAC ROM'.format(e, m))
+                result = 'FAIL'
+        print('TEST {}!'.format(result))
+
+    def start(self):
+        self.get_if_and_mac_list()
+        self.read_mac_from_nvmem()
+        self.compare_eth_mac_with_macrom()
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--segment', '-S',
+                        help='Segment number of PCIe device')
+    parser.add_argument('--bus', '-B',
+                        help='Bus number of PCIe device')
+    parser.add_argument('--device', '-D',
+                        help='Device number of PCIe device')
+    parser.add_argument('--function', '-F',
+                        help='Function number of PCIe device')
+    parser.add_argument('--offset',
+                        default='0',
+                        help='read mac address from a offset address')
+    args, left = parser.parse_known_args()
+
+    args = convert_argument_str2hex(
+        args, ['segment', 'bus', 'device', 'function', 'offset'])
+
+    f = FpgaFinder(args.segment, args.bus, args.device, args.function)
+    devs = f.find()
+    for d in devs:
+        print(
+            'bdf: {segment:04x}:{bus:02x}:{dev:02x}.{func:x}'.format(
+                **d))
```

Comment:
For this unpacking, should it be single * instead of double **? Also, can this line be formatted consistently like the others.

Reply:
fixed